File: ai-study-planner/validate_plan.js
const fs = require("node:fs");
const path = require("node:path");

const projectDir = __dirname;
const subjects = JSON.parse(fs.readFileSync(path.join(projectDir, "subjects.json"), "utf8"));
const plan = JSON.parse(fs.readFileSync(path.join(projectDir, "plan.json"), "utf8"));

const expectedTopics = new Set();
const plannedTopics = new Set();

for (const subject of subjects) {
  for (const topic of subject.topics) {
    expectedTopics.add(`${subject.subject}::${topic}`);
  }
}

for (const day of plan) {
  for (const task of day.tasks) {
    plannedTopics.add(`${task.subject}::${task.topic}`);
  }
}

const missing = [...expectedTopics].filter(topic => !plannedTopics.has(topic));
const extra = [...plannedTopics].filter(topic => !expectedTopics.has(topic));

if (missing.length || extra.length) {
  console.log("Validace selhala.");
  if (missing.length) console.log("Chybi temata:", missing.join(", "));
  if (extra.length) console.log("Prebyvaji temata:", extra.join(", "));
  process.exit(1);
}

console.log("Validace prosla: vsechna temata ze subjects.json jsou v plan.json.");

File: ai-study-planner/validate_plan.py
from __future__ import annotations

import json
from pathlib import Path


PROJECT_DIR = Path(__file__).parent
SUBJECTS_FILE = PROJECT_DIR / "subjects.json"
PLAN_FILE = PROJECT_DIR / "plan.json"


def load_json(path: Path):
    with path.open("r", encoding="utf-8") as file:
        return json.load(file)


def main() -> None:
    subjects = load_json(SUBJECTS_FILE)
    plan = load_json(PLAN_FILE)

    expected_topics = {
        (subject["subject"], topic)
        for subject in subjects
        for topic in subject["topics"]
    }
    planned_topics = {
        (task["subject"], task["topic"])
        for day in plan
        for task in day["tasks"]
    }

    missing = expected_topics - planned_topics
    extra = planned_topics - expected_topics

    if missing or extra:
        print("Validace selhala.")
        if missing:
            print("Chybi temata:", sorted(missing))
        if extra:
            print("Prebyvaji temata:", sorted(extra))
        raise SystemExit(1)

    print("Validace prosla: vsechna temata ze subjects.json jsou v plan.json.")


if __name__ == "__main__":
    main()
